File: expense-report/typescript/tests/ExpenseItemBuilder.ts
import { Category } from '../src/Category.js';
import { ExpenseItem } from '../src/ExpenseItem.js';
import { Money } from '../src/Money.js';

export class ExpenseItemBuilder {
  private _description = 'Office supplies';
  private _amount = 25;
  private _category: Category = Category.Other;

  withDescription(description: string): this { this._description = description; return this; }
  withAmount(amount: number): this { this._amount = amount; return this; }
  withCategory(category: Category): this { this._category = category; return this; }

  asMeal(amount = 30): this {
    return this.withCategory(Category.Meals).withAmount(amount).withDescription('Team lunch');
  }

  asTravel(amount = 200): this {
    return this.withCategory(Category.Travel).withAmount(amount).withDescription('Flight');
  }

  asAccommodation(amount = 150): this {
    return this.withCategory(Category.Accommodation).withAmount(amount).withDescription('Hotel stay');
  }

  asEquipment(amount = 800): this {
    return this.withCategory(Category.Equipment).withAmount(amount).withDescription('Laptop');
  }

  build(): ExpenseItem {
    return new ExpenseItem(this._description, new Money(this._amount), this._category);
  }
}

File: expense-report/python/tests/expense_item_builder.py
from __future__ import annotations

from expense_report import Category, ExpenseItem, Money


class ExpenseItemBuilder:
    def __init__(self) -> None:
        self._description = "Office supplies"
        self._amount = 25.0
        self._category = Category.OTHER

    def with_description(self, description: str) -> ExpenseItemBuilder:
        self._description = description
        return self

    def with_amount(self, amount: float) -> ExpenseItemBuilder:
        self._amount = amount
        return self

    def with_category(self, category: Category) -> ExpenseItemBuilder:
        self._category = category
        return self

    def as_meal(self, amount: float = 30.0) -> ExpenseItemBuilder:
        return self.with_category(Category.MEALS).with_amount(amount).with_description("Team lunch")

    def as_travel(self, amount: float = 200.0) -> ExpenseItemBuilder:
        return self.with_category(Category.TRAVEL).with_amount(amount).with_description("Flight")

    def as_accommodation(self, amount: float = 150.0) -> ExpenseItemBuilder:
        return self.with_category(Category.ACCOMMODATION).with_amount(amount).with_description("Hotel stay")

    def as_equipment(self, amount: float = 800.0) -> ExpenseItemBuilder:
        return self.with_category(Category.EQUIPMENT).with_amount(amount).with_description("Laptop")

    def build(self) -> ExpenseItem:
        return ExpenseItem(self._description, Money(self._amount), self._category)
